src/EditorFeatures/Core/Implementation/IntelliSense/AsyncCompletion/CompletionSource.cs

```diff
@@ -254,29 +351,44 @@ static void CheckForExperimentStatus(ITextView textView, in CompletionOptions op
             {
                 AsyncCompletionLogger.LogExpanderUsage();
 
+                // It's possible we didn't provide expanded items at the beginning of completion session because it was slow even if the feature is enabled.
+                // ExpandedItemsTask would be available in this case, so we just need to return its result.
+                // We are doing this in a spinning fashion because the lifetime of expandedItemsTask is associated with the completion session, we don't
+                // want to cancel that task when this "GetExpandedCompletionContextAsync" operation is being canceled, so we will still be able to retrieve
+                // the results later.
+                if (session.Properties.TryGetProperty<Task<(AsyncCompletionData.CompletionContext, CompletionList)>>(ExpandedItemsTask, out var expandedItemsTask))
+                {
+                    // Make sure the task is removed when returning expanded items,
+                    // so duplicated items won't be added in subsequent list updates.
+                    session.Properties.RemoveProperty(ExpandedItemsTask);
```

Comment:
does anything run concurrently with this?  any risk that something else might still see this prior to us removing this code if it's running concurrently with us?

Reply:
@AmadeusW confirmed that calls to Roslyn are synchronized using a worker thread
https://github.com/dotnet/roslyn/pull/59045#issuecomment-1024774747